### PR TITLE
fix(CI): fix msrv checks on github run on push to main

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR introduces changes to improve the CI workflow in the repository. The main goal is to separate the logic for pushes to the main branch from pull requests.
1. Created a new workflow file, `CI-push.yml`, specifically for pushes to the main branch.
2. Modified the existing CI.yml file to trigger only on pull requests.
3. Removed the logic that checked for differences in crates in the `CI-push.yml` file.
The logic we had in place for checking differences in crates in a pull request is not valid when we push directly to the main branch.

The `CI-push.yml` file contains a single job, `push-to-main`, which runs on Ubuntu latest and performs the following steps:
- Checks out the repository.
- Installs Rust and cargo-hack.
- Executes the `cargo hack check` command to check all workspace packages without considering specific differences.

By separating the workflows, we can now ensure that CI builds on the main branch do not fail due to the logic designed for PRs. The `CI-push.yml` workflow specifically targets pushes to the main branch and performs the necessary cargo-hack checks on all workspace packages.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by lsampras — APPROVED** [review]

File: .github/workflows/CI-pr.yml [hunk 1]
```diff
@@ -193,7 +184,7 @@ jobs:
       - name: Cargo hack redis_interface
         if: env.redis_interface_changes_exist == 'true'
         shell: bash
-        run: cargo hack check --each-feature --no-dev-deps -p redis_interface
+        run: cargo hack check --each-feature --no-dev-deps
```
Comment (lsampras) [inline .github/workflows/CI-pr.yml:187]: will need to use package specific change here

**Review by lsampras — CHANGES_REQUESTED** [review]

File: .github/workflows/CI-pr.yml [hunk 2]
```diff
@@ -84,8 +74,9 @@ jobs:
         uses: actions/checkout@v3
 
       - name: "Fetch base branch"
-        shell: bash
-        run: git fetch origin $GITHUB_BASE_REF --depth 1
+        run: |
+          git config submodule.recurse true
```
Comment (lsampras) [inline .github/workflows/CI-pr.yml:78]:
```suggestion
```

File: .github/workflows/CI-pr.yml [hunk 3]
```diff
@@ -252,8 +243,9 @@ jobs:
...
```
Comment (lsampras) [inline .github/workflows/CI-pr.yml:247]:
```suggestion
```

**Review by lsampras — CHANGES_REQUESTED** [review]

File: .github/workflows/CI-pr.yml [hunk 4]
```diff
@@ -252,8 +243,9 @@ jobs:
...
+          git fetch origin $GITHUB_BASE_REF --depth 1
```
Comment (lsampras) [inline .github/workflows/CI-pr.yml:248]:
```suggestion
        run: git fetch origin $GITHUB_BASE_REF --depth 1
```

File: .github/workflows/CI-pr.yml [hunk 5]
```diff
@@ -84,8 +74,9 @@ jobs:
...
```
Comment (lsampras) [inline .github/workflows/CI-pr.yml:79]:
```suggestion
        run: git fetch origin $GITHUB_BASE_REF --depth 1
```

**Review by lsampras — CHANGES_REQUESTED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Comment by lsampras** [comment]
@ShankarSinghC Can we also move `merge_group` trigger to `CI-push.yaml`

**Pull request merged**